cmd/s3-peer-client.go

```diff
@@ -19,125 +19,74 @@ package cmd
+				loginMethod: "S3.LoginHandler",
+			}
 
-func (s3p *s3Peers) GetPeerClient(peer string) *AuthRPCClient {
-	// Take a read lock
-	s3p.mutex.RLock()
-	defer s3p.mutex.RUnlock()
-	return s3p.rpcClients[peer]
-}
+			bmsc[ep.Host] = &remoteBMS{newAuthClient(&cfg)}
 
-// Initializes a new RPC connection (or closes and re-opens if it
-// already exists) to a peer. Note that peer address is in `host:port`
-// format.
-func (s3p *s3Peers) InitS3PeerClient(peer string) {
-	// Take a write lock
-	s3p.mutex.Lock()
-	defer s3p.mutex.Unlock()
-
-	if s3p.rpcClients[peer] != nil {
-		_ = s3p.rpcClients[peer].Close()
-		delete(s3p.rpcClients, peer)
+			peers = append(peers, ep.Host)
+		}
 	}
-	authCfg := &authConfig{
-		accessKey:   serverConfig.GetCredential().AccessKeyID,
-		secretKey:   serverConfig.GetCredential().SecretAccessKey,
-		address:     peer,
-		secureConn:  isSSL(),
-		path:        path.Join(reservedBucket, s3Path),
-		loginMethod: "S3.LoginHandler",
+
+	return s3Peers{
+		bmsClients: bmsc,
+		IsDistXL:   len(peers) > 1,
+		Peers:      peers,
 	}
-	s3p.rpcClients[peer] = newAuthClient(authCfg)
 }
 
-func (s3p *s3Peers) Close() error {
-	// Take a write lock
-	s3p.mutex.Lock()
-	defer s3p.mutex.Unlock()
-
-	for _, v := range s3p.rpcClients {
-		if err := v.Close(); err != nil {
-			return err
-		}
-	}
-	s3p.rpcClients = nil
-	s3p.peers = nil
-	return nil
+func initGlobalS3Peers(eps []*url.URL) {
+	globalS3Peers = mkPeersFromEndpoints(eps)
 }
 
-// Returns the network addresses of all Minio servers in the cluster in `host:port` format.
-func getAllPeers(eps []*url.URL) (peers []string) {
-	if eps == nil {
-		return nil
-	}
-	peers = []string{globalMinioAddr} // Starts with a default peer.
-	for _, ep := range eps {
-		if ep == nil {
-			return nil
-		}
-		// Rest of the peers configured.
-		peers = append(peers, ep.Host)
-	}
-	return peers
+func (s3p *s3Peers) GetPeerClient(peer string) BucketMetaState {
+	return s3p.bmsClients[peer]
 }
 
-// Make RPC calls with the given method and arguments to all the given
-// peers (in parallel), and collects the results. Since the methods
-// intended for use here, have only a success or failure response, we
-// do not return/inspect the `reply` parameter in the RPC call. The
-// function attempts to connect to a peer only once, and returns a map
-// of peer address to error response. If the error is nil, it means
-// the RPC succeeded.
-func (s3p *s3Peers) SendRPC(peers []string, method string, args interface {
-	SetToken(token string)
-	SetTimestamp(tstamp time.Time)
-}) map[string]error {
+// Send bucket metadata updates to all given nodes. The update calls
+// are sent in parallel, and errors are returned per node in a
+// map. The updates are sent to a type implementing the
+// BucketMetaState interface. This makes sure that the local node is
+// directly updated, and remote nodes are updated via RPC calls.
+func (s3p *s3Peers) SendUpdate(peers []string, args interface{}) map[string]error {
```

Comment:
Same reason as the other review comment for `SendUpdate` (appears below this comment for me).